core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -258,26 +262,60 @@ private[spark] class CoarseGrainedExecutorBackend(
     System.exit(code)
   }
 
-  private def decommissionSelf(): Boolean = {
-    logInfo("Decommissioning self w/sync")
-    try {
-      decommissioned = true
-      // Tell master we are are decommissioned so it stops trying to schedule us
-      if (driver.nonEmpty) {
-        driver.get.askSync[Boolean](DecommissionExecutor(executorId))
+  private def shutdownIfDone(): Unit = {
+    val numRunningTasks = executor.numRunningTasks
+    logInfo(s"Checking to see if we can shutdown have ${numRunningTasks} running tasks.")
+    if (executor.numRunningTasks == 0) {
+      if (env.conf.get(STORAGE_DECOMMISSION_ENABLED)) {
+        val allBlocksMigrated = env.blockManager.decommissionManager match {
+          case Some(m) => m.allBlocksMigrated
+          case None => false // We haven't started migrations yet.
+        }
+        if (allBlocksMigrated) {
+          logInfo("No running tasks, all blocks migrated, stopping.")
+          exitExecutor(0, "Finished decommissioning", notifyDriver = true)
```

Comment:
`exitExecutor` asynchronously sends RemoveExecutor to the driver. Does that actually make it to the driver ? There is also this question about if we should be using the same `Shutdown`/`StopExecutor` codepath for doing the stopping ? (But althought it seems that we do want to intimate to the driver that the executor is being removed). 

Interestingly, the driver does indeed respond back with a `StopExecutor` and does trigger the clean shutdown path in the executor, but again I wonder if it is too late for it. Perhaps we shouldn't be calling `System.exit` here ?

Also, as currently written, this `exitExecutor` could cause job failures: Since the `TaskSchedulerImpl` will treat the `ExecutorLossReason` send by the executor to the driver as an `exitCausedByApp` and thus penalize the task. Instead, I think we shouldn't penalize the running job on a planned executor decommission. One workaround might be to actually respond back to the driver with `ExecutorDecommission` (which is not used elsewhere currently) and then handle that specifically in the `TaskSchedulerImpl`'s determination of `exitCausedByApp`.

Reply:
So it's my understanding the `TaskSchedulerImpl` shouldn't have any job failures because we've waited for all the tasks on the executor to finish before calling this code path. Unless is there something I've missed there?

I think swapping out exit executor for instead telling the driver to stop the executor and avoiding the `system.exit` makes sense either way though.

Reply:
I was talking about the case where we get shot down before we had a chance to cleanly exit on line 276. Say for example, some time out expires and the executor/node is brought down. 

Are `decom.sh` and `decommission-slave.sh` expected to wait until the executor/worker process has properly shut down ? I think they have some timeouts in them to kill the executor ? Or consider a spot kill scenario where you got some warning (like 2 minutes) and then the machine is yanked out. 

In this case, the executor will eventually be marked loss via a heartbeat/timeout. And that loss would be deemed as the fault of the task, and could cause job failures. I am wondering if we can fix that scenario of an unclean exit ? 

One workaround I suggested above was to send a message to the driver saying that the executor is going to go away soon. When that happens (in a clean or unclean way), that loss shouldn't be attributed to the task. 

Perhaps this unclean executor loss/timeout handling is follow up work ? We (or rather I) can create Jira's for this under the parent ticket :-).

Reply:
Sure, although I _think_ this behaviour is covered by the changes in https://github.com/apache/spark/pull/26440/files (we only increment failures if the executors previous state was not decommissioning).

Reply:
Can you please double check that ? I couldn't find this behavior when scouring TaskSchedulerImpl,  and TaskSetManager. The only place we check for an executor being decommissioned in that PR is when scheduling tasks (in CoarseGrainedSchedulerBackend#isExecutorActive). Thanks !

Reply:
Can you point to where in TaskSchedulerImpl it's going to fail the job? `core/src/main/scala/org/apache/spark/deploy/master/Master.scala` is where the current code is, but there might be an additional case that needs to be covered.

Reply:
https://github.com/apache/spark/blob/a2c055715ccf2992e399cef3768b1299c24d9a82/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L980

In this match block, we will hit the default case which will treat the failure as having been caused by the app and thus penalize it. 

This routine is called from `TaskScheduler.executorLost`

Reply:
Thanks :) If you want to make a PR for that I'd be happy to review/merge since I think that would not depend on any of the in-flight PRs just the current code in master.

Reply:
Absolutely !. Thanks